polkadot/xcm/xcm-builder/src/asset_exchange/single_asset_adapter/adapter.rs

```diff
@@ -0,0 +1,209 @@
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Single asset exchange adapter.
+
+use core::marker::PhantomData;
+use frame_support::{ensure, traits::tokens::fungibles};
+use pallet_asset_conversion::{QuotePrice, SwapCredit};
+use sp_std::vec;
+use xcm::prelude::*;
+use xcm_executor::{
+	traits::{AssetExchange, MatchesFungibles},
+	AssetsInHolding,
+};
+
+/// An adapter from [`pallet_asset_conversion::SwapCredit`] and
+/// [`pallet_asset_conversion::QuotePrice`] to [`xcm_executor::traits::AssetExchange`].
+///
+/// This adapter takes just one fungible asset in `give` and allows only one fungible asset in
+/// `want`. If you need to handle more assets in either `give` or `want`, then you should use
+/// another type that implements [`xcm_executor::traits::AssetExchange`] or build your own.
+///
+/// `exchange_asset` will return an error if there's more than one asset in `want`.
+pub struct SingleAssetExchangeAdapter<AssetConversion, Fungibles, Matcher, AccountId>(
+	PhantomData<(AssetConversion, Fungibles, Matcher, AccountId)>,
+);
+impl<AssetConversion, Fungibles, Matcher, AccountId> AssetExchange
+	for SingleAssetExchangeAdapter<AssetConversion, Fungibles, Matcher, AccountId>
+where
+	AssetConversion: SwapCredit<
+			AccountId,
+			Balance = u128,
+			AssetKind = Fungibles::AssetId,
+			Credit = fungibles::Credit<AccountId, Fungibles>,
+		> + QuotePrice<Balance = u128, AssetKind = Fungibles::AssetId>,
+	Fungibles: fungibles::Balanced<AccountId, Balance = u128>,
+	Matcher: MatchesFungibles<Fungibles::AssetId, Fungibles::Balance>,
+{
+	fn exchange_asset(
+		_: Option<&Location>,
+		give: AssetsInHolding,
+		want: &Assets,
+		maximal: bool,
+	) -> Result<AssetsInHolding, AssetsInHolding> {
+		let mut give_iter = give.fungible_assets_iter();
+		let give_asset = give_iter.next().ok_or_else(|| {
+			log::trace!(
+				target: "xcm::SingleAssetExchangeAdapter::exchange_asset",
+				"No fungible asset was in `give`.",
+			);
+			give.clone()
+		})?;
+		ensure!(give_iter.next().is_none(), give.clone()); // We only support 1 asset in `give`.
+		ensure!(want.len() == 1, give.clone()); // We only support 1 asset in `want`.
+		let want_asset = if let Some(asset) = want.get(0) {
+			asset
+		} else {
+			log::trace!(
+				target: "xcm::SingleAssetExchangeAdapter::exchange_asset",
+				"No asset was in `want`.",
+			);
```

Comment:
I guess this trace will never be triggered due to the `ensure` statement before.